### PR TITLE
ASG-2921: New functions on db to support retrieving tokens.

### PR DESCRIPTION
**Pull request by bosvos**

### CODE REVIEW

File: auth.go [hunk 1]
```diff
@@ -996,7 +1004,54 @@ func (x *Central) GetAuthenticatorIdentities(getIdentitiesFlag GetIdentitiesFlag
 func (x *Central) GetRoleGroupDB() RoleGroupDB {
 	return x.roleGroupDB
 }
+func RemoveGroupFromAllUsers(c *Central, groupIDString string) error {
```
Comment (FritzOnFire) [inline auth.go:1007]:
```suggestion
func (c *Central) RemoveGroupFromAllUsers(groupIDString string) error {
```

Comment (FritzOnFire) [inline auth.go:1007]: Please leave an empty line between functions

File: auth.go [hunk 2]
```diff
@@ -996,7 +1004,54 @@ func (x *Central) GetAuthenticatorIdentities(getIdentitiesFlag GetIdentitiesFlag
...
+
+		// Encode the permit
+		if modified {
+			perm.Roles = EncodePermit(groups)
+			fmt.Printf("Removing group %v: setting permit for %v\n", groupID, user.Email)
+
+			//Set the permit
+			err = c.SetPermit(user.UserId, perm)
+			if err != nil {
+				fmt.Printf("Error: %v\n", err)
```
Comment (FritzOnFire) [inline auth.go:1047]: I think we should remove this line, as the caller of this function can log this if it feels like it.

File: db.go [hunk 3]
```diff
@@ -381,6 +383,15 @@ func (x *dummySessionDB) Read(sessionkey string) (*Token, error) {
 	return token, nil
 }
 
+func (x *dummySessionDB) GetAllTokens(includeExpired bool) ([]*Token, error) {
+	//TODO implement me
+	panic("implement me")
+}
+func (x *dummySessionDB) GetAllOAuthTokenIDs() ([]string, error) {
```
Comment (FritzOnFire) [inline db.go:390]: Please leave an empty line between functions.

File: auth.go [hunk 4]
```diff
@@ -996,7 +1004,54 @@ func (x *Central) GetAuthenticatorIdentities(getIdentitiesFlag GetIdentitiesFlag
...
+			if g == groupID {
+				groups = append(groups[:i], groups[i+1:]...)
+				modified = true
+			}
+		}
+
+		// Encode the permit
+		if modified {
+			perm.Roles = EncodePermit(groups)
+			fmt.Printf("Removing group %v: setting permit for %v\n", groupID, user.Email)
```
Comment (FritzOnFire) [inline auth.go:1042]: I would rather make use of `c.Log.Debugf`. But Infof might also be fine, as I don't see something like this filling up a log file, or standard out.

File: auth.go [hunk 5]
```diff
@@ -996,7 +1004,54 @@ func (x *Central) GetAuthenticatorIdentities(getIdentitiesFlag GetIdentitiesFlag
...
+			if err != nil {
+				fmt.Printf("Error: %v\n", err)
+				return err
+			}
+		}
+	}
+
+	return nil
+}
 func (x *Central) IsShuttingDown() bool {
```
Comment (FritzOnFire) [inline auth.go:1055]: Please leave an empty space between functions.

File: db_sql.go [hunk 6]
```diff
@@ -648,28 +649,97 @@ func (x *sqlSessionDB) Read(sessionkey string) (*Token, error) {
...
+		} else {
+			token.Identity = user.Email
+		}
+		if oauthID.Valid {
+			token.OAuthSessionID = oauthID.String
+		}
+
+		return token, nil, true
+	}
+	return nil, nil, false
```
Comment (FritzOnFire) [inline db_sql.go:683]: If I am understanding this code correctly, then I would remove the `else` and just have it's content be the rest of the function.

File: db_sql.go [hunk 7]
```diff
@@ -648,28 +649,97 @@ func (x *sqlSessionDB) Read(sessionkey string) (*Token, error) {
...
-			token.Username = user.Username
-			token.Email = user.Email
-			if user.Type == UserTypeLDAP {
-				token.Identity = user.Username
+			return nil, nil
+		}
+	}
+}
+
+func (x *sqlSessionDB) populateToken(token *Token, epermit string, userId int64, oauthID sql.NullString) (*Token, error, bool) {
```
Comment (FritzOnFire) [inline db_sql.go:661]: I might be wrong... but I don't think you need the `bool` at the end. At least, not with the code that I can see in this PR.

File: db_sql.go [hunk 8]
```diff
@@ -648,28 +649,97 @@ func (x *sqlSessionDB) Read(sessionkey string) (*Token, error) {
 			return nil, err
 		}
 	} else {
-		if err := token.Permit.Deserialize(epermit); err != nil {
-			return nil, err
+		t, err2, done := x.populateToken(token, epermit, userId, oauthID)
```
Comment (FritzOnFire) [inline db_sql.go:652]: The current implementation of `populateToken` will always return true for `done`. I think you should rather check the `err`, like the previous code did.

**Review by FritzOnFire — CHANGES_REQUESTED** [review]
Quite a few changes, but it should simplify the code a bit.

File: auth.go (hunk 4 above)
Comment (bosvos) [reply auth.go:1055]: Keeping it as Infof for now. We still look at the auth logfile very regularly to check if groups have been recognized and appropriate action has been taken on the user.

File: db_sql.go (hunk 7 above)
Comment (bosvos) [reply db_sql.go:676]: Removed.

**Pull request closed without merging**